drivers/sdhc/sdhc_renesas_ra.c

```diff
@@ -0,0 +1,736 @@
+
+	k_sem_give(&priv->thread_lock);
+
+	return 0;
+}
+
+/*
+ * Set SDHC io properties
+ */
+static int sdhc_ra_set_io(const struct device *dev, struct sdhc_io *ios)
+{
+	struct sdhc_ra_priv *priv = dev->data;
+	const struct sdhc_ra_config *cfg = dev->config;
+	struct st_sdmmc_instance_ctrl *p_ctrl = &priv->sdmmc_ctrl;
+	int fsp_err;
+	int ret = 0;
+
+	uint8_t bus_width;
+	uint32_t bus_width_reg;
+
+	if (ios->bus_width > 0) {
+		bus_width_reg = 0;
+		/* Set bus width, SD bus interface doesn't support 8BIT */
+		switch (ios->bus_width) {
+		case SDHC_BUS_WIDTH1BIT:
+			bus_width = 1;
+			bus_width_reg = 4;
+			break;
+		case SDHC_BUS_WIDTH4BIT:
+			bus_width = 4;
+			break;
+		default:
+			ret = -ENOTSUP;
+			goto end;
+		}
+
+		if (priv->bus_width != bus_width) {
+			/* Set the bus width in the SDHI peripheral. */
+			((R_SDHI0_Type *)cfg->regs)->SD_OPTION =
+				SDHI_PRV_SD_OPTION_DEFAULT |
+				(bus_width_reg << SDHI_PRV_SD_OPTION_WIDTH8_BIT);
+			priv->bus_width = bus_width;
+		}
+	}
+
+	if (ios->clock) {
+		if (ios->clock > priv->props.f_max || ios->clock < priv->props.f_min) {
+			LOG_ERR("Proposed clock outside supported host range");
+			return -EINVAL;
+		}
+
+		if (priv->bus_clock != (uint32_t)ios->clock) {
+			fsp_err = r_sdhi_max_clock_rate_set(p_ctrl, ios->clock);
+			ret = err_fsp2zep(fsp_err);
+			if (ret < 0) {
+				goto end;
+			}
+			priv->bus_clock = ios->clock;
+		}
+	}
+
+	if (ios->timing > 0) {
+		/* Set I/O timing */
+		if (priv->timing != ios->timing) {
+			switch (ios->timing) {
+			case SDHC_TIMING_LEGACY:
+			case SDHC_TIMING_HS:
+			case SDHC_TIMING_SDR12:
+			case SDHC_TIMING_SDR25:
+				break;
+			default:
+				LOG_ERR("Timing mode not supported for this device");
+				ret = -ENOTSUP;
+				break;
+			}
+
+			priv->timing = ios->timing;
+		}
+	}
+end:
+
+	return ret;
+}
+
+/*
+ * Get host properties
+ */
+static int sdhc_ra_get_host_props(const struct device *dev, struct sdhc_host_props *props)
+{
+	struct sdhc_ra_priv *priv = dev->data;
+
+	memcpy(props, &priv->props, sizeof(struct sdhc_host_props));
+	return 0;
+}
+
+static int sdhc_ra_init(const struct device *dev)
+{
+	const struct sdhc_ra_config *config = dev->config;
+	struct sdhc_ra_priv *priv = dev->data;
+	fsp_err_t fsp_err;
+	int timeout = SDHI_PRV_ACCESS_TIMEOUT_US;
+	int ret = 0;
+
+	sdmmc_event[priv->channel].transfer_completed = false;
+	k_sem_init(&sdmmc_event[priv->channel].transfer_sem, 1, 1);
+
+	/* Configure dt provided device signals when available */
+	ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		return ret;
+	}
+	if (priv->sdhi_en.port != NULL) {
+		int err = gpio_pin_configure_dt(&priv->sdhi_en, GPIO_OUTPUT_HIGH);
+
+		if (err) {
+			return err;
+		}
+		k_sleep(K_MSEC(50));
+	}
+
+	k_sem_init(&priv->thread_lock, 1, 1);
+	fsp_err = R_SDHI_Open(&priv->sdmmc_ctrl, &priv->fsp_config);
+	ret = err_fsp2zep(fsp_err);
+
+	if (ret < 0) {
+		LOG_INF("\r\nR_SDHI_Open error: %d\r\n", fsp_err);
```

Comment:
Don't need the `\r` here, and likely no leading `\n` either

Reply:
Thank you. I removed it